File: smartmoneyserver/src/main/java/fpt/aptech/server/enums/date/DateRange.java
package fpt.aptech.server.enums.date;

/**
 * Định nghĩa các khoảng thời gian tương đối mà API báo cáo hỗ trợ.
 * <p>
 * Dùng làm tham số `range` trong các API GET, giúp Frontend không cần
 * tự tính toán `startDate` và `endDate` phức tạp.
 */
public enum DateRange {
    // === TUẦN ===
    THIS_WEEK,      // Tuần này (từ Thứ Hai đến Chủ Nhật)
    LAST_WEEK,      // Tuần trước

    // === THÁNG ===
    THIS_MONTH,     // Tháng này
    LAST_MONTH,     // Tháng trước

    // === QUÝ ===
    THIS_QUARTER,   // Quý này
    LAST_QUARTER,   // Quý trước

    // === NĂM ===
    THIS_YEAR,      // Năm nay
    LAST_YEAR,      // Năm trước

    // === TƯƠNG LAI (Dùng chung cho tất cả) ===
    FUTURE,         // Tương lai (từ ngày mai trở đi vô tận)

    /**
     * Trường hợp đặc biệt: Người dùng tự chọn một khoảng ngày bất kỳ.
     * Khi API nhận được `range=CUSTOM`, nó sẽ sử dụng `startDate` và `endDate`
     * được gửi kèm để xác định khoảng thời gian.
     */
    CUSTOM;
}

File: APP MOBILE/smart_money/lib/core/enums/date_range.dart
/// Các khoảng thời gian tương đối: tuần này, tháng này, năm nay, v.v.
enum DateRange {
  thisWeek,
  lastWeek,
  thisMonth,
  lastMonth,
  thisQuarter,
  lastQuarter,
  thisYear,
  lastYear,
  future,
  custom;

  static DateRange fromString(String value) {
    final map = {
      'THIS_WEEK': DateRange.thisWeek,
      'LAST_WEEK': DateRange.lastWeek,
      'THIS_MONTH': DateRange.thisMonth,
      'LAST_MONTH': DateRange.lastMonth,
      'THIS_QUARTER': DateRange.thisQuarter,
      'LAST_QUARTER': DateRange.lastQuarter,
      'THIS_YEAR': DateRange.thisYear,
      'LAST_YEAR': DateRange.lastYear,
      'FUTURE': DateRange.future,
      'CUSTOM': DateRange.custom,
    };
    return map[value] ?? DateRange.custom;
  }

  String get apiValue {
    final map = {
      DateRange.thisWeek: 'THIS_WEEK',
      DateRange.lastWeek: 'LAST_WEEK',
      DateRange.thisMonth: 'THIS_MONTH',
      DateRange.lastMonth: 'LAST_MONTH',
      DateRange.thisQuarter: 'THIS_QUARTER',
      DateRange.lastQuarter: 'LAST_QUARTER',
      DateRange.thisYear: 'THIS_YEAR',
      DateRange.lastYear: 'LAST_YEAR',
      DateRange.future: 'FUTURE',
      DateRange.custom: 'CUSTOM',
    };
    return map[this] ?? 'CUSTOM';
  }

  String get displayName {
    switch (this) {
      case DateRange.thisWeek:
        return 'This Week';
      case DateRange.lastWeek:
        return 'Last Week';
      case DateRange.thisMonth:
        return 'This Month';
      case DateRange.lastMonth:
        return 'Last Month';
      case DateRange.thisQuarter:
        return 'This Quarter';
      case DateRange.lastQuarter:
        return 'Last Quarter';
      case DateRange.thisYear:
        return 'This Year';
      case DateRange.lastYear:
        return 'Last Year';
      case DateRange.future:
        return 'Future';
      case DateRange.custom:
        return 'Custom';
    }
  }
}

